bser_parse.c

```diff
@@ -0,0 +1,374 @@
+#include <stdio.h>
+#include <string.h>
+
+#include "bser.h"
+
+/**
+ * BSER parsing is done lazily, and does not copy strings out of the buffer.
+ * This is accomplished by initializing each new object that we encounter
+ * as an "unparsed" object, which contains only a pointer to the data buffer
+ * (which itself has a cursor).  The first time an object is queried or a
+ * value is requested of it, it accesses the buffer and reads only the data
+ * it needs to satisfy the request.  For unit types, this usually means reading
+ * the tag and all the data.  For strings, the length is read and a pointer
+ * to the string data in the buffer is stored in the object.
+ *
+ * Parsing occurs in the order that the data appears in the buffer.  That is,
+ * if an object's data appears earlier in the buffer than a different object,
+ * then the first object is guaranteed to be parsed first.  The buffer is a
+ * stream and cannot be randomly accessed.  The in-order parsing happens in
+ * the implementation of the 'array' and 'object' types, which are the only
+ * composite types in the system.  When these are initially parsed, the read
+ * only the header information (including the size) and allocate an array to
+ * use for storage and this storage is initialized to all unparsed object.
+ * However, when an element or a field is accessed, the query methods ensures
+ * that all elements or fields before the requested field are fully-parsed
+ * (from first to last).
+ */
+
+static bser_t* new_unparsed(bser_buffer_t* buffer, bser_t* fill)
+{
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_UNPARSED;
+    fill->value.unparsed.buffer = buffer;
+    return fill;
+}
+
+static bser_t* new_error(const char* message, bser_t* fill)
+{
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_ERROR;
+    fill->value.error_message = message;
+    return fill;
+}
+
+static bser_t* new_compact_object(
+    bser_buffer_t* buffer, bser_t* header, bser_t* fill)
+{
+    assert(bser_is_array(header));
+    size_t sz = bser_array_size(header);
+    bser_key_value_pair_t* fields = malloc(sizeof(*fields) * sz);
+    if (fields == NULL) {
+        new_error("Could not allocate enough memory to hold "
```

Comment:
But you still need to return here because otherwise you're going to write to a NULL pointer.